File: backend/services/density_optimizer.py
MAX_BULLET_WORDS = 15
MAX_BULLETS_PER_SLIDE = 4
MAX_TITLE_WORDS = 10

def truncate_text(text, max_words):
    if not isinstance(text, str):
        return text
    words = text.split()
    if len(words) > max_words:
        return " ".join(words[:max_words]) + "..."
    return text

def optimize_presentation(slides):
    optimized_slides = []
    
    for slide in slides:
        # 1. Truncate title
        if "title" in slide:
            slide["title"] = truncate_text(slide["title"], MAX_TITLE_WORDS)
            
        # 2. Pagination for informational slides
        category = slide.get("slide_category", "")
        
        # Fallback if category is missing but slide_type matches typical informational slides
        if not category and slide.get("slide_type") in ["standard_text", "two_column_image", "comparison", "summary_takeaways"]:
            category = "informational"
            
        if category == "informational" and "bullets" in slide and isinstance(slide["bullets"], list):
            # Truncate bullets
            processed_bullets = [truncate_text(b, MAX_BULLET_WORDS) for b in slide["bullets"]]
            
            if len(processed_bullets) <= MAX_BULLETS_PER_SLIDE:
                slide["bullets"] = processed_bullets
                optimized_slides.append(slide)
            else:
                # Chunking
                for i in range(0, len(processed_bullets), MAX_BULLETS_PER_SLIDE):
                    chunk = processed_bullets[i:i + MAX_BULLETS_PER_SLIDE]
                    new_slide = slide.copy()
                    new_slide["bullets"] = chunk
                    if i > 0:
                        new_title = new_slide.get("title", "")
                        if not new_title.endswith("(Cont.)"):
                            new_slide["title"] = f"{new_title} (Cont.)"
                    optimized_slides.append(new_slide)
        else:
            # Still truncate bullets if present, just in case
            if "bullets" in slide and isinstance(slide["bullets"], list):
                slide["bullets"] = [truncate_text(b, MAX_BULLET_WORDS) for b in slide["bullets"]]
            optimized_slides.append(slide)
            
    return optimized_slides

File: backend/services/densityOptimizer.js
/**
 * Density Optimizer for PptxGenJS
 * Ported from density_optimizer.py but with relaxed limits.
 */

// Relaxed limits so we don't aggressively chop AI output
const MAX_TITLE_WORDS = 20; // Increased from 10
const MAX_BULLET_WORDS = 30; // Increased from 15
const MAX_BULLETS_PER_SLIDE = 6; // Increased from 4

function truncateText(text, maxWords) {
    if (!text || typeof text !== 'string') return text;
    const words = text.split(' ');
    if (words.length <= maxWords) return text;
    return words.slice(0, maxWords).join(' ') + '...';
}

function truncateSlide(slide) {
    const s = { ...slide };
    if (s.title) s.title = truncateText(s.title, MAX_TITLE_WORDS);
    if (s.subtitle) s.subtitle = truncateText(s.subtitle, MAX_TITLE_WORDS);
    if (s.bullets && Array.isArray(s.bullets)) {
        s.bullets = s.bullets.map(b => truncateText(b, MAX_BULLET_WORDS));
    }
    // Also truncate cards/columns
    if (s.column_left?.bullets) {
        s.column_left.bullets = s.column_left.bullets.map(b => truncateText(b, MAX_BULLET_WORDS));
    }
    if (s.column_right?.bullets) {
        s.column_right.bullets = s.column_right.bullets.map(b => truncateText(b, MAX_BULLET_WORDS));
    }
    if (s.cards && Array.isArray(s.cards)) {
        s.cards = s.cards.map(c => ({
            ...c,
            card_title: truncateText(c.card_title, MAX_TITLE_WORDS),
            card_text: truncateText(c.card_text, MAX_BULLET_WORDS)
        }));
    }
    if (s.items && Array.isArray(s.items)) {
        s.items = s.items.map(c => ({
            ...c,
            item_title: truncateText(c.item_title, MAX_TITLE_WORDS),
            item_text: truncateText(c.item_text, MAX_BULLET_WORDS)
        }));
    }
    return s;
}

/**
 * Optimizes the presentation data by truncating long text.
 * No longer splits slides into multiple slides to preserve intended layout flow.
 */
function optimizePresentation(slides) {
    if (!Array.isArray(slides)) return [];
    const optimized = [];
    
    for (const slide of slides) {
        const s = truncateSlide(slide);
        
        // We will just cap bullets to MAX_BULLETS_PER_SLIDE, but not split
        if (s.bullets && Array.isArray(s.bullets)) {
            if (s.bullets.length > MAX_BULLETS_PER_SLIDE) {
                s.bullets = s.bullets.slice(0, MAX_BULLETS_PER_SLIDE);
            }
        }
        
        optimized.push(s);
    }
    return optimized;
}

module.exports = {
    optimizePresentation
};
